src/oscar/apps/partner/abstract_models.py

```diff
@@ -189,11 +191,27 @@ def allocate(self, quantity):
 
         This normally happens when a product is bought at checkout.  When the
         product is actually shipped, then we 'consume' the allocation.
+
         """
+        # Atomic update
+        (self.__class__.objects
+            .filter(pk=self.pk)
+            .update(num_allocated=(
+                Coalesce(F('num_allocated'), Value(0)) + quantity)))
+
+        # Make sure the current object is up-to-date
         if self.num_allocated is None:
             self.num_allocated = 0
         self.num_allocated += quantity
```

Comment:
Sorry if I wasn't clear. I think we mean the same thing. With your code, the database will have the correct value, but if there is a race condition, wouldn't `self.num_allocated` still contain the "faulty" quantity?